### ISSUE TITLE
Slow Download Speed and Frequent Retry Errors in geedim

### ISSUE THREAD

**Issue by mht2953658596** [leftfield-geospatial/geedim#26]
My colleagues and I have been using the geedim library to download data, and we've encountered some issues that didn't exist before. It seems like the library has recently faced some restrictions:

Slow Download Speed: The data download process has become extremely slow, which wasn't an issue previously.
Frequent Retry Errors: We are now seeing messages like "Tile downloaded failed, retry 3 of 5" very frequently. This is a new problem that didn't occur before.
It appears that the library might be facing some recent limitations. Has anything changed recently that could be causing these issues? Are there any solutions or workarounds available?

Thank you for your assistance!

![image](https://github.com/user-attachments/assets/36a6a9f4-e8bb-4236-bd59-45766734cca0)
We have confirmed that the issue is not related to our network

![屏幕截图 2024-08-28 223424](https://github.com/user-attachments/assets/01cbd4e4-27ab-40d6-8aa8-e4372a4d1aa3)

**Comment by mht2953658596** [comment]
```
def calculate_spm_by_gee(m: str, shp_path: str, output_dir: str):
    roi = ee.FeatureCollection(shapefile.Reader(shp_path).__geo_interface__)

    def rmCloud(image):
        cloudShadowBitMask = (1 << 4)
        cloudBitMask = (1 << 3)
        qa = image.select("QA_PIXEL")
        mask = (qa.bitwiseAnd(cloudShadowBitMask).eq(0)
                .And(qa.bitwiseAnd(cloudBitMask).eq(0)))
        return image.updateMask(mask)

    def apply_scale_factors57(image):
        optical_bands = image.select('SR_B.').multiply(0.0000275).add(-0.2)
        thermal_bands = image.select('ST_B6').multiply(0.00341802).add(149.0)
        return image.addBands(optical_bands, None, True).addBands(thermal_bands, None, True)

    def apply_scale_factors89(image):
        optical_bands = image.select('SR_B.').multiply(0.0000275).add(-0.2)
        thermal_bands = image.select('ST_B.*').multiply(0.00341802).add(149.0)
        return image.addBands(optical_bands, None, True).addBands(thermal_bands, None, True)

    def addNDWI57(image):
        ndwi = image.normalizedDifference(['SR_B2', 'SR_B4']).rename('addNDWI57')
        return ndwi

    def addNDWI89(image):
        ndwi = image.normalizedDifference(['SR_B3', 'SR_B5']).rename('addNDWI89')
        return ndwi

    if int(m) < 2012:
        l5 = (ee.ImageCollection("LANDSAT/LT05/C02/T1_L2")
              .filterDate(f"{m}-01-01", f"{m}-12-31")
              .filterBounds(roi)
              .filter(ee.Filter.lte('CLOUD_COVER', 30))
              .map(rmCloud)
              .map(apply_scale_factors57)
              .mean()
              .select(['SR_B1', 'SR_B2', 'SR_B3', 'SR_B4']))

        ndwi = addNDWI57(l5)
        water = ndwi.gt(-0.005).updateMask(ndwi.lt(10))
        water = water.updateMask(water.gt(0.5))

        l5elicloud = ee.ImageCollection("LANDSAT/LT05/C02/T1_L2").map(rmCloud).map(apply_scale_factors57)
        mean_ecloud = l5elicloud.mean().clip(roi)
        SPM = mean_ecloud.expression(
            '0.0430 + 45 * red', {
                'red': mean_ecloud.select('SR_B3'),
            }).float()
        mean_ecloud = mean_ecloud.addBands(SPM.rename("SPM"))
        SPM = ee.Image(water).multiply(SPM).toFloat()

    elif int(m) == 2012:
        l7 = (ee.ImageCollection("LANDSAT/LE07/C02/T1_L2")
              .filterDate(f"{m}-01-01", f"{m}-12-31")
              .filterBounds(roi)
              .filter(ee.Filter.lte('CLOUD_COVER', 30))
              .map(rmCloud)
              .map(apply_scale_factors57)
              .mean()
              .select(['SR_B1', 'SR_B2', 'SR_B3', 'SR_B4']))

        ndwi = addNDWI57(l7)
        water = ndwi.gt(-0.005).updateMask(ndwi.lt(10))
        water = water.updateMask(water.gt(0.5))

        l7elicloud = ee.ImageCollection("LANDSAT/LE07/C02/T1_L2").map(rmCloud).map(apply_scale_factors57)
        mean_ecloud = l7elicloud.mean().clip(roi)
        SPM = mean_ecloud.expression(
            '0.0430 + 45 * red', {
                'red': mean_ecloud.select('SR_B3'),
            }).float()
        mean_ecloud = mean_ecloud.addBands(SPM.rename("SPM"))
        SPM = ee.Image(water).multiply(SPM).toFloat()

    elif int(m) > 2012:
        l8 = (ee.ImageCollection("LANDSAT/LC08/C02/T1_L2")
              .filterDate(f"{m}-01-01", f"{m}-12-31")
              .filterBounds(roi)
              .filter(ee.Filter.lte('CLOUD_COVER', 30))
              .map(rmCloud)
              .map(apply_scale_factors89)
              .mean()
              .select(['SR_B2', 'SR_B3', 'SR_B4', 'SR_B5']))

        ndwi = addNDWI89(l8)
        water = ndwi.gt(-0.005).updateMask(ndwi.lt(10))
        water = water.updateMask(water.gt(0.5))

        l8elicloud = ee.ImageCollection("LANDSAT/LC08/C02/T1_L2").map(rmCloud).map(apply_scale_factors89)
        mean_ecloud = l8elicloud.mean().clip(roi)
        SPM = mean_ecloud.expression(
            '0.0430 + 45 * red', {
                'red': mean_ecloud.select('SR_B4'),
            }).float()
        mean_ecloud = mean_ecloud.addBands(SPM.rename("SPM"))
        SPM = ee.Image(water).multiply(SPM).toFloat()

    imagename = "SD_" + m + "_SPM" + ".tif"
    filename = os.path.join(output_dir, imagename)
    geemap.download_ee_image(SPM, filename, scale=30, crs='EPSG:4326', max_tile_size=4, region=roi.geometry())

```

**Comment by dugalh** [comment]
Hello.  I was able to download an "SPM" image using your code without any retries or other problems.  I used a 1 degree by 2 degree region.  Other download tests I've tried are working too.  I suspect the problem is related to your network.  It is more network reliability than bandwidth that would cause problems like these.

**Comment by mht2953658596** [comment]
@dugalh  Unfortunately, I've tried several network nodes, but still exist this error

**Comment by dugalh** [comment]
You can try setting the [``num_threads`` argument](https://geedim.readthedocs.io/en/latest/_generated/geedim.mask.MaskedImage.download.html#geedim-mask-maskedimage-download) to a small value (e.g. 4) to reduce the number of concurrent requests.  The default value is based on the number of CPU cores, with a maximum of 32.  

Also see the notes here if you are not using a cloud project: https://developers.google.com/earth-engine/guides/usage#legacy_quota_decrease.  The current Earth Engine concurrent request limit for legacy projects (20) is less than the geedim maximum (32), so could be the cause of your problem.

**Comment by mht2953658596** [comment]
@dugalh Thank you for your reply. First, I removed the toFloat() in the line SPM = ee.Image(water).multiply(SPM).toFloat() in the code, and after that, the download process started. I'm not sure why, but previously it was stuck at 0% the entire time. I also followed your suggestion to reduce the number of concurrent threads and specified the project during the initialization. The issue has improved; although the download is slower, at least it's now working.
